feat(Forms): add `country` prop to Field.PostalCodeAndCity to support other countries than Norway

joakbjerk opened the pull request:
- [x] Add example
- [x] Document props
- [x] Add tests
- [x] Add path support for `country` prop
- [x] Figure out wonky behavior in new tests 🤔
- [x] Fix `mask` type error
- [x] Add demo for path based country values
- [x] Update Postal Code input width in example

boyum commented on packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/PostalCodeAndCity/demos.mdx:
```diff
@@ -44,6 +44,12 @@ By using the `itemPath` property, you can iterate over an array and use the `pos
 
 <Examples.ValidationRequired />
 
+### Non Norwegain Postal Codes
```
```suggestion
### Non-Norwegian Postal Codes
```

boyum commented on packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/PostalCodeAndCity/demos.mdx:
```diff
@@ -44,6 +44,12 @@ By using the `itemPath` property, you can iterate over an array and use the `pos
 
+### Non Norwegain Postal Codes
+
+If you want to allow for a postal code that is not Norwegain, just set the `country` property to the desired country and add your own custom validation.
```
```suggestion
If you want to allow for a postal code that is not Norwegian, just set the `country` property to the desired country and add your own custom validation.
```

tujoworker commented:
> How to fix? refresh cache? Close the CLI and browser and restart these app and processes did work for me when I saw such an error.

tujoworker commented on packages/dnb-eufemia/src/extensions/forms/Field/PostalCodeAndCity/PostalCodeAndCity.tsx:
```diff
@@ -1,26 +1,67 @@
     city = {},
     help,
     width = 'large',
+    country = 'no',
```
We have a file called `defaults.ts`. But we don't have a country default there. Maybe we can/should add:

```ts
export const COUNTRY = 'NO' // ISO 3166-1 alpha-2
```

tujoworker commented on packages/dnb-eufemia/src/extensions/forms/Value/PostalCodeAndCity/PostalCodeAndCity.tsx:
```diff
@@ -40,5 +33,17 @@ function PostalCodeAndCity(props: Props) {
   return <StringValue {...stringValueProps} />
 }
 
+function usePostalCodeAndCityValue(props: Props) {
+  let value = props.value ?? undefined
+
+  const valueProps = useValueProps(props)
+
+  if (valueProps.value) {
+    value = valueProps.value
+  }
+
+  return value
+}
```
```suggestion
function usePostalCodeAndCityValue(props: Props) {
  return useValueProps(props).value ?? props.value
}
```

This should be equivalent with less code.

langz commented:
Looks good, I'll do a quick round of testing after lunch 🐔

langz commented on packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/PostalCodeAndCity/demos.mdx:
```diff
@@ -44,6 +44,18 @@ By using the `itemPath` property, you can iterate over an array and use the `pos
 
+### Path Based Country
+
+The `country` prop supports a field path as value. This allows you to set the `country` based on the value of another field.
```
```suggestion
The `country` property supports a field path as value. This allows you to set the `country` based on the value of another field.
```

langz commented on packages/dnb-eufemia/src/extensions/forms/Field/PostalCodeAndCity/PostalCodeAndCity.tsx:
```diff
@@ -1,26 +1,71 @@
+        : isNorway
+        ? [/\d/, /\d/, /\d/, /\d/]
+        : undefined,
+      pattern: postalCode.pattern
```
```suggestion
      pattern: postalCode?.pattern
```

langz replied:
```suggestion
      pattern: postalCode?.pattern
```

langz commented:
I added a few minor commits, which is just "code styling", and could be discussed if it improves the code-readability or not 😂 

Anyhow, I like this, and it works good :) 

# I have a few thoughts

## In the given example(Non-Norwegian Postal Codes) why would I as a consumer want to set the country?
<img width="506" alt="Screenshot 2024-10-16 at 12 38 21" src="https://github.com/user-attachments/assets/3fcecccc-3e07-4ceb-b34d-5ef0f97d9eac">

## The spacing between the input fields in the component(PostalCodeAndCity) when displaying country is a bit tight/crammed vertically. Should it be so? It perhaps makes sense that it's a bit tighter vertically spacing here, compared to if it would just be three separate fields(not put together in a component with a given purpose):
<img width="416" alt="Screenshot 2024-10-16 at 12 39 14" src="https://github.com/user-attachments/assets/ecf6411c-1ef3-4d33-b339-be93c3c4c7bd">

joakbjerk commented:
> I added a few minor commits, which is just "code styling", and could be discussed if it improves the code-readability or not 😂
> 
> Anyhow, I like this, and it works good :)
> ### I have a few thoughts
> #### In the given example(Non-Norwegian Postal Codes) why would I as a consumer want to set the country to be "de"?
> <img alt="Screenshot 2024-10-16 at 12 38 21" width="506" src="https://private-user-images.githubusercontent.com/1359205/377015326-3fcecccc-3e07-4ceb-b34d-5ef0f97d9eac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwNzU2OTMsIm5iZiI6MTcyOTA3NTM5MywicGF0aCI6Ii8xMzU5MjA1LzM3NzAxNTMyNi0zZmNlY2NjYy0zZTA3LTRjZWItYjM0ZC01ZWYwZjk3ZDllYWMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTZUMTA0MzEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDNmNzNhZmM1Zjc1ZjY4NTQxZjI4MjI1NzQ1NWI2ZWFhYWE4OWViZmFmZjc1OTM0OGQxZmMwNTQ0NWU0NjRmZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.GkZOVKvfS1DfLueMILmTVENpBy1Vfmd1COIRtiTcz4c">
> #### The spacing between the input fields in the component(PostalCodeAndCity) when displaying country is a bit tight/crammed vertically. Should it be so? It perhaps makes sense that it's a bit tighter vertically spacing here, compared to if it would just be three separate fields(not put together in a component with a given purpose), but it still feels a bit tight:
> <img alt="Screenshot 2024-10-16 at 12 39 14" width="416" src="https://private-user-images.githubusercontent.com/1359205/377015995-ecf6411c-1ef3-4d33-b339-be93c3c4c7bd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwNzU2OTMsIm5iZiI6MTcyOTA3NTM5MywicGF0aCI6Ii8xMzU5MjA1LzM3NzAxNTk5NS1lY2Y2NDExYy0xZWYzLTRkMzMtYjMzOS1iZTkzYzNjNGM3YmQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTZUMTA0MzEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDUwMmYyNzM0Yjk1ODQ2ZDc5MTkwMWQzNGRiZDdjNjU4ZWE4MDMwMmFlZjRjZTViYzE0MTMyYzBiZmU5NmFlYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.zFiKTnCPQ1OQLAmzupAvXBf0o51ZGEcud8z7UyBq4Hk">

Added space between the fields in demo 👍

The `country` property thingy stems from [this slack thread](https://dnb-it.slack.com/archives/C04NL73S2Q5/p1728472603329999), where it was discussed that setting the country prop to something other than `NO` would remove the default Norwegian validation used by the component. But as you say, it's not clear why one would to that, as the property really just functions as a kinda of unclear `removeDefaultValidation` property, since we do not provide any built in validation rules based on the value of the country prop itself 🤷‍♂️

tujoworker commented:
> The spacing between the input fields in the component(PostalCodeAndCity) when displaying country is a bit tight/crammed vertically. Should it be so?

Maybe there's a more declarative way of doing this. But I think, but having the country prop, it allows us eventually to progressively add support for other countries. But it could be done via a "plugin" system that a dev can import and provide for certain countries, just to lower the amount of extra code it is needed to support many other countries. 

But the most important thing to support is removing the limitation – when "myCountryPath" has changed.

langz commented on packages/dnb-eufemia/src/extensions/forms/Field/PostalCodeAndCity/PostalCodeAndCity.tsx:
```diff
@@ -41,30 +41,31 @@ function PostalCodeAndCity(props: Props) {
+  } = city
+
+  const {
+    mask: postalCodeMask = isNorway ? [/\d/, /\d/, /\d/, /\d/] : undefined,
```
Comment for reference

tujoworker commented:
For some reason the placeholder does not update:
<img width="445" alt="Screenshot 2024-10-16 at 13 25 16" src="https://github.com/user-attachments/assets/a7fbadf0-e079-44c9-aff5-51a3d3186e23">

Also, the size is too large. We should consider to change the default size. Just a view pixels are missing:

<img width="112" alt="Screenshot 2024-10-16 at 13 23 16" src="https://github.com/user-attachments/assets/908d0f65-2d67-4255-80f0-786e5312a596">

tujoworker commented on packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/PostalCodeAndCity/Examples.tsx:
```diff
@@ -172,3 +172,35 @@ export const IterateArray = () => {
+        <Field.SelectCountry
+          path="/myCountry"
+          defaultValue="NO"
+          bottom="medium"
```
```suggestion
```

We should use Flex.Stack or a card with stack: `<Card stack>`

joakbjerk commented:
> For some reason the placeholder does not update: <img alt="Screenshot 2024-10-16 at 13 25 16" width="445" src="https://private-user-images.githubusercontent.com/1501870/377030472-a7fbadf0-e079-44c9-aff5-51a3d3186e23.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwOTkwODIsIm5iZiI6MTcyOTA5ODc4MiwicGF0aCI6Ii8xNTAxODcwLzM3NzAzMDQ3Mi1hN2ZiYWRmMC1lMDc5LTQ0YzktYWZmNS01MWEzZDMxODZlMjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTZUMTcxMzAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmE1NzIyZGJmMzExNDY3ZDRjMjg1NDQxZDg0ZDUyZGU2NzZjNGE2NmJiZmJmMDU4YmY5YjVkNTdmYzI5NWFiYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.r8A8nnLrdPF25qvQeMmy1TJ4sfr-JyDfthVecYEF_U4">
> 
> Also, the size is too large. We should consider to change the default size. Just a view pixels are missing:
> <img alt="Screenshot 2024-10-16 at 13 23 16" width="112" src="https://private-user-images.githubusercontent.com/1501870/377030288-908d0f65-2d67-4255-80f0-786e5312a596.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwOTkwODIsIm5iZiI6MTcyOTA5ODc4MiwicGF0aCI6Ii8xNTAxODcwLzM3NzAzMDI4OC05MDhkMGY2NS0yZDY3LTQyNTUtODBmMC03ODZlNTMxMmE1OTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTZUMTcxMzAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjQ3OTNhMmU4ZmUzMDVjMzA0YzA3OWExOTNjOTc1MjlhNjM3NTdkNTcwOTQzMTY1NmVjZWJmMDFlNzlhMGU1MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.pwPjQahFT6eFtCjaOqYxAgQk5PbeIftMn-feGV0Vq0M">

Should be working again now ⭐️

joakbjerk commented:
> > For some reason the placeholder does not update: <img alt="Screenshot 2024-10-16 at 13 25 16" width="445" src="https://private-user-images.githubusercontent.com/1501870/377030472-a7fbadf0-e079-44c9-aff5-51a3d3186e23.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwOTkwODIsIm5iZiI6MTcyOTA5ODc4MiwicGF0aCI6Ii8xNTAxODcwLzM3NzAzMDQ3Mi1hN2ZiYWRmMC1lMDc5LTQ0YzktYWZmNS01MWEzZDMxODZlMjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTZUMTcxMzAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmE1NzIyZGJmMzExNDY3ZDRjMjg1NDQxZDg0ZDUyZGU2NzZjNGE2NmJiZmJmMDU4YmY5YjVkNTdmYzI5NWFiYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.r8A8nnLrdPF25qvQeMmy1TJ4sfr-JyDfthVecYEF_U4">
> > Also, the size is too large. We should consider to change the default size. Just a view pixels are missing:
> > <img alt="Screenshot 2024-10-16 at 13 23 16" width="112" src="https://private-user-images.githubusercontent.com/1501870/377030288-908d0f65-2d67-4255-80f0-786e5312a596.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwOTkwODIsIm5iZiI6MTcyOTA5ODc4MiwicGF0aCI6Ii8xNTAxODcwLzM3NzAzMDI4OC05MDhkMGY2NS0yZDY3LTQyNTUtODBmMC03ODZlNTMxMmE1OTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTZUMTcxMzAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjQ3OTNhMmU4ZmUzMDVjMzA0YzA3OWExOTNjOTc1MjlhNjM3NTdkNTcwOTQzMTY1NmVjZWJmMDFlNzlhMGU1MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.pwPjQahFT6eFtCjaOqYxAgQk5PbeIftMn-feGV0Vq0M">

Should be working again now ⭐️

joakbjerk commented:
> For some reason the placeholder does not update: <img alt="Screenshot 2024-10-16 at 13 25 16" width="445" src="https://private-user-images.githubusercontent.com/1501870/377030472-a7fbadf0-e079-44c9-aff5-51a3d3186e23.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwOTk1NTgsIm5iZiI6MTcyOTA5OTI1OCwicGF0aCI6Ii8xNTAxODcwLzM3NzAzMDQ3Mi1hN2ZiYWRmMC1lMDc5LTQ0YzktYWZmNS01MWEzZDMxODZlMjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTZUMTcyMDU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjRkMjE2OWY0MDc0ZTVkZDc3ZmRiZTNmNzlmNjA0NmZkMTRmZmRiOTcyNmZjOTZkNmIxNWJhMjg0MjU2MWVkMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.0Bez4RMX3Cj5k8fzaFXXjb5_igylRf4fPdWwJGCPAxw">
> 
> Also, the size is too large. We should consider to change the default size. Just a few pixels are missing:
> <img alt="Screenshot 2024-10-16 at 13 23 16" width="112" src="https://private-user-images.githubusercontent.com/1501870/377030288-908d0f65-2d67-4255-80f0-786e5312a596.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwOTk1NTgsIm5iZiI6MTcyOTA5OTI1OCwicGF0aCI6Ii8xNTAxODcwLzM3NzAzMDI4OC05MDhkMGY2NS0yZDY3LTQyNTUtODBmMC03ODZlNTMxMmE1OTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTZUMTcyMDU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDA3ZGEyMTkzMzVlMGJlOTgxZmYyMmQzMWIwZTgzZTk2MWVmZDc2MzlkNzg5NzQ2NDcwMTM3YjljMmRiMmNkNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.A9XZoa1GDnuUS4UciiZbwJkrl0SX9ys3VG_anHcVvqY">

Regarding the width of the input, I agree, it's to wide, but do we have an option to make width on the input accommodate the mask, or placeholder length? I've tried setting the size prop to the mask length, like you could with the masked input, without any luck ☹️

tujoworker commented:
Added this PR #4142 which we can use to set a larger width on the postal code field.

joakbjerk commented:
With the new width property:

<img width="860" alt="Screenshot 2024-10-18 at 10 21 54" src="https://github.com/user-attachments/assets/637a3273-4db9-4a11-96c3-3c8e365defb7">

langz reviewed (approved):
Looks great 🚀

tujoworker commented:
Nice work 🥁

tujoworker commented:
My last commit changes how we handle the defaults. I think, this way we are more flexible to adjust specific needs in the future. Also, for most other countries, the size will be larger of the postal code field.

<img width="458" alt="Screenshot 2024-10-18 at 11 10 49" src="https://github.com/user-attachments/assets/e7fa7b26-5ebf-41a4-80c8-259972122580">

tujoworker commented:
:tada: This PR is included in version 10.53.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@dnb/eufemia/v/10.53.0)
- [GitHub release](https://github.com/dnbexperience/eufemia/releases/tag/v10.53.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
